### PR TITLE
Prompt for signature on refresh [Fixes #220]

### LINKED ISSUE #220: Prompt signing when refreshing page if user has already connected wallet
https://www.notion.so/efdn/Prompt-signing-when-user-refreshes-page-016b181e32ac458fafba7f0d2349916d

Currently if the user has connected their wallet in the past, when they refresh or revisit the page it appears as though they haven't and that they need to reconnect their wallet. This is largely a side effect of MACI, and that we need to have the user re-sign when connecting to the application. Since we are able to know that the user has used this application, but just need this signature, we should prompt for the signature when the application is loaded for users that have connected in the past.

### PR DESCRIPTION
**Pull request by wackerow**
### Changes
- Adds new `localStorage.ts` utility functions to manage localStorage, and enable storage of chosen wallet type.
- Does not store any sensitive data in this unencrypted form
- Within `$web3` plugin, now stores the chosen wallet in localStorage on login and removes the chosen wallet on disconnect
- This also checks if there is an existing choice on page load, and if so, triggers a signature/login
- Adds `user` key to `$web3` export for storage, instead of returning the User object from method
- Moves the `loginUser` functionality to `App.vue` from `WalletModal.vue`, and decorates the function with `@Watch('$web3.user')` which will run this function whenever the stores user changes, returning early if no user found.
- While working in `App.vue`: Added functionality to clear the intervals that are set in the `created()` method when the component unmounts to prevent memory leaks. 
- Updates wallet login message

### tl;dr:
User will now be prompted for a signature on page refresh if they've previously logged in using a chosen wallet provider. Upon "disconnecting" this choice is removed from localStorage, and in this case refreshing the page will not prompt user.

### CODE REVIEW

File: vue-app/src/utils/connection.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    } catch (e) {
+      return defaultValue
+    }
+  }
+  return defaultValue
+}
+
+export function lsRemove(key: string) {
+  return localStorage.removeItem(lsGetKey(key))
+}
```
Comment (pettinarip) [inline vue-app/src/utils/connection.ts]: Everthing in here is related to local storage manipulation. I would call this file `localStorage.ts`.

File: vue-app/src/utils/connection.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import pkg from '@/../package.json'
+
+function lsGetKey(key: string) {
+  return `${pkg.name}.${key}`
```
Comment (pettinarip) [inline vue-app/src/utils/connection.ts]: cool!

File: vue-app/src/plugins/Web3/index.ts [hunk 3]
```diff
@@ -111,6 +120,15 @@ export default {
       }
 
       plugin.provider = null
+      plugin.user = null
+    }
+
+    // If previous provider was found, initiate connection.
+    if (alreadyConnectedProvider) {
+      const response = plugin.connectWallet(alreadyConnectedProvider)
+      if (response.state !== 'fulfilled') {
```
Comment (pettinarip) [inline vue-app/src/plugins/Web3/index.ts]: Wondering what `state` does in this case since we are not `await`ing the promise. I mean, in which other cases are we going to face a different value than `'fulfilled'`?

**Comment by pettinarip** [comment]
Looks great @wackerow. I left a couple of small comments.

I'm wondering if we should consider the following case:
1- User connects with MM
2- Refreshes the page
3- Rejects the signature
4- Tries to connect with a different wallet, WalletConnect
5- Should we manually disconnect from the previous wallet (MM)? or doesn't it hurt us to leave it as connected?

File: vue-app/src/plugins/Web3/index.ts (hunk 3 above)
Comment (wackerow) [reply vue-app/src/plugins/Web3/index.ts]: @pettinarip So this only returns `'fulfilled'` is the user actually signs... Adding this basically is checking if the user cancels the prompt. If they do, it will then clear the localStorage.

ie: If the user clicks "cancel" on MetaMask -> returns `response.state = 'pending'`... 

When I tried to wrap this in `await` syntax, the app does not respond and update with the logged in user. Clicking "Connect" and signing just leads to the same screen with the "Connect" button still showing. Keeping it synchronous as above seems to work properly.

**Comment by wackerow** [comment]
@pettinarip See the other comment about this, as rejecting the signature should basically start someone back at square one, with no wallet preference saved. Clicking connect at this point will offer the MetaMask or WalletConnect option again.

File: vue-app/src/plugins/Web3/index.ts (hunk 3 above)
Comment (pettinarip) [reply vue-app/src/plugins/Web3/index.ts]: I just tested this and understood what you want to do.

Actually, checking the promise `state` prop is not going to tell you if the user has accepted or not the prompt. That state is part of the Promise lifecycle (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). When the Promise is `resolve`d or `reject`ed, the state variable is going to change.

So, in this case, `response.state` is always going to be `pending` because we are not waiting for it to end. Basically, the condition is always going to be true.

Regarding what you want to achieve here I think we could do the following:
```ts
if (alreadyConnectedProvider) {
  // What we really want to do is to clean the `alreadyConnectedProvider` provider, not do a complete disconnection.
  // This is mainly to avoid future problems if we add more logic in the `disconnectWallet` method
  lsRemove(CONNECTED_PROVIDER) 
  plugin.connectWallet(alreadyConnectedProvider)
}
```

**Comment by pettinarip** [comment]
> @pettinarip See the other comment about this, as rejecting the signature should basically start someone back at square one, with no wallet preference saved. Clicking connect at this point will offer the MetaMask or WalletConnect option again.

Got it. Yes, looks great that approach. To remove the `alreadyConnectedProvider` from the ls before forcing the connection again.

File: vue-app/src/App.vue [hunk 4]
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
     await this.$store.dispatch(LOAD_RECIPIENT_REGISTRY_INFO)
   }
 
+  beforeDestroy() {
```
Comment (samajammin) [inline vue-app/src/App.vue:106]: 👍

File: vue-app/src/App.vue [hunk 5]
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
...
+    for (const interval of Object.keys(this.intervals)) {
+      clearInterval(this.intervals[interval])
+    }
+  }
+
+  @Watch('$web3.user')
```
Comment (samajammin) [inline vue-app/src/App.vue:112]: So is this how class based components adds watchers?

File: vue-app/src/App.vue [hunk 6]
```diff
@@ -96,6 +103,28 @@ export default class App extends Vue {
...
+  loginUser = async () => {
+    if (!this.$web3.user) return
+
+    this.$store.commit(SET_CURRENT_USER, this.$web3.user)
```
Comment (samajammin) [inline vue-app/src/App.vue:116]: Won't this `SET_CURRENT_USER` overwrite all other user data that $web3 doesn't have (such as `isRegistered` & `brightId`)?

Comment (samajammin) [reply vue-app/src/App.vue:116]: I suppose that's ok since we subsequently dispatch `LOAD_USER_INFO`? Is that the thinking here?

File: vue-app/src/utils/localStorage.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import pkg from '@/../package.json'
+
+function lsGetKey(key: string) {
+  return `${pkg.name}.${key}`
```
Comment (samajammin) [inline vue-app/src/utils/localStorage.ts:4]: Could you explain why we use `pkg.name` - just to give our keys a namespace?

File: vue-app/src/plugins/Web3/index.ts (hunk 3 above)
Comment (samajammin) [reply vue-app/src/plugins/Web3/index.ts]: Agreed - thanks for the explanation @pettinarip!

**Review by samajammin — CHANGES_REQUESTED** [review]
This looks amazing! 🤩 

Agree w/ @pettinarip's commet re: the `alreadyConnectedProvider` check - let's make that change before merging.

**Comment by samajammin** [comment]
Tried it on both MetaMask & WalletConnect - it's looking super slick!

File: vue-app/src/App.vue (hunk 5 above)
Comment (wackerow) [reply vue-app/src/App.vue:112]: https://github.com/kaorun343/vue-property-decorator#Watch 👍🏻

File: vue-app/src/App.vue (hunk 6 above)
Comment (pettinarip) [reply vue-app/src/App.vue:116]: Well, you are right but we shouldn't reach the case where we overwrite something by mistake. Mainly because this function is only going to be called when `$web3.user` changes. That happens in two moments:

1. When `disconnectWallet` is called: the top guard clause is for this case.
2. When `connectWallet` is called: we are going to overwrite the user but we are calling the rest of the data that completes the user state.

**Review by samajammin — APPROVED** [review]

File: vue-app/src/utils/localStorage.ts (hunk 7 above)
Comment (wackerow) [reply vue-app/src/utils/localStorage.ts:4]: Yes, this was just to ensure a unique namespace
